components/script/dom/globalscope.rs

```diff
@@ -2561,26 +2570,58 @@ impl GlobalScope {
                 let ar = enter_realm(&*self);
 
                 let _aes = AutoEntryScript::new(self);
-                let options =
-                    unsafe { CompileOptionsWrapper::new(*cx, filename.as_ptr(), line_number) };
-
-                debug!("evaluating Dom string");
-                let result = unsafe {
-                    Evaluate2(
-                        *cx,
-                        options.ptr,
-                        &mut transform_str_to_source_text(code),
-                        rval,
-                    )
-                };
 
-                if !result {
-                    debug!("error evaluating Dom string");
-                    unsafe { report_pending_exception(*cx, true, InRealm::Entered(&ar)) };
-                }
+                unsafe {
+                    let options = CompileOptionsWrapper::new(*cx, filename.as_ptr(), line_number);
+
+                    debug!("compiling Dom string");
+                    rooted!(in(*cx) let compiled_script =
+                        Compile1(
+                            *cx,
+                            options.ptr,
+                            &mut transform_str_to_source_text(code),
+                        )
+                    );
+
+                    if compiled_script.is_null() {
+                        debug!("error compiling Dom string");
+                        report_pending_exception(*cx, true, InRealm::Entered(&ar));
+                        return false;
+                    }
+
+                    // When `ScriptPrivate` for the compiled script is undefined,
+                    // we need to set it so that it can be used in dynamic import context.
+                    if GetScriptPrivate(*compiled_script).is_undefined() {
+                        let base_url = script_url.unwrap_or(self.api_base_url());
+
+                        debug!("Set script private for {}", base_url);
+
+                        let module_script_data = Box::new(ModuleScript::new(
+                            base_url,
+                            ScriptFetchOptions::default_classic_script(&self),
+                            // We can't initialize an module owner here because
+                            // the executing context of script might be different
+                            // from the dynamic import script's executing context.
+                            None,
+                        ));
+
+                        SetScriptPrivate(
+                            *compiled_script,
+                            &PrivateValue(Box::into_raw(module_script_data) as *const _),
```

Comment:
After reading the full PR, we do not call Box::from_raw. We will need to use [SetScriptPrivateReferenceHooks](https://searchfox.org/mozilla-central/rev/89814940895946b48b4c04c702efd2c676ec8e7e/js/src/jsapi.h#1793-1798) - I believe if we use Rc instead of Box (and Rc::into_raw and Rc::from_raw), we can call clone() on the value in the addref hook, and drop the value in the release hook, and it should work ok. See https://searchfox.org/mozilla-central/rev/89814940895946b48b4c04c702efd2c676ec8e7e/dom/script/LoadedScript.cpp#70-88 for Gecko's implementation.

Reply:
Thanks for commenting on this! I planned to ask about it but forgot >_<

To make sure I understand correctly,

With the functionality of reference counting, when we use `Rc` instead of `Box` and then call `clone()` in `addref` hook, the counting will be added automatically, is that correct?

and a question here 👇 

1. For dropping values in release hook, I'm not clear how can I do that? Should I explicitly call `drop(private_reference)`?

btw, maybe it's good to handle it in this PR? I imagine the private references won't be released if we don't do the drop in the release hook; we're making all scripts (both classic and module) to own an private reference so it might be good to release them properly?

Reply:
Yes, if we do `let val = RC::from_raw(...); mem::forget(val.clone()); mem::forget(val);` in the addref hook, that will add one reference to the value. In the release hook, we do `let _val = Rc::from_raw(...);` and it will automatically subtract one reference when we return.

Reply:
Fixed in 731a16e!

I basically copied the code in previous comment to add ref hook but I'm not sure if I clearly understand how the `mem::forget` works here.

Does that mean

```rust
let val = Rc::from_raw(value); // 1. Get the existing script reference from Rc raw pointer
mem::forget(val.clone());        // 2. Increment the Rc count by doing a `clone` but release it without running drop
mem::forget(val);                    // 3. Release the existing value without running drop
```

With `mem::forget`, we can release the memory here but still make it counted for the Rc pointer; is that correct?